### DISCUSSION TITLE
[PROPOSAL] Top Level Extensions Methods

### DISCUSSION THREAD

**Discussion by tebeco** [dotnet/csharplang#5993] [Language Ideas]
(@ maintaines, Feel free to edit / re-formulate / fix typos if you spot anything that seems obvious to you)

Hello,

I was wondering if, that now we have `Top Level Statement`, that would be usable for `Top Level Extensions Methods`

## Reason(s):
99% of the time, as a developper you never type `BarExtensions.Do(bar)` and the class name is "often" `<Type>Extensions`

## Syntax
As bad as it could be, this is what I had in mind:
Instead of writting:
```cs
namespace Foo;

public static class BarExtensions
{
  public static A Do(this Bar bar) {}
}
```
Prefer:
```cs
namespace Foo;

public static A Do(this Bar bar) {}
```

## Restrictions
Not sure all makes sense, but it could have restriction to help:
That file should only ever contains extensions method
  ```cs
  namespace Foo;

  public static A Do(this Bar bar) {}
  public static A Doo(Bar bar) {}  // INVALID: is not an extension method (could be added later)
  ```

  ```cs
  namespace Foo;

  public static A Do(this Bar bar) {}
  public record A(string B) // INVALID: nothing else than extension in that file (could be added later)
  ```

### Naive approach
Is it possible to use the detected extended type and suffix it with `Extensions`:
```cs
public static A Do(this Bar bar) {} // Bar is detected as extended type
```
=>
```cs
public static class BarExtensions
{
  public static A Do(this Bar bar) {}
}
```

* First try to see if `public static class <ExtendedType>Extensions` conflict with an existing symbol => use that
* On conflict
  * Should it break compilation ?
  * Should it try to generate a unique name at compile time ? (see `Cons` section below)

### Multiple type extensions in same file
```cs
namespace Foo;

public static A Do(this Bar bar);
public static A Do(this Kix kix);
```
Should they emitted in same class or should it try to be consistent in `BarExtensions` and `KixExtensions` ?
What is `BarExtensions` is possible but `KixExtensions` fails ? etc ...
(many question where I'm likely the wrong person to have answers, it's just issue I could think about)

## Cons / open questions
Assuming conflicts:
### Consumer
* Same solution:
  for whatever reason decide to use `BarExtensions.Do(bar)` the generated name can be hard to find
  => one would still use the "existing" syntax
* Nuget consumer
  this one is "meh" I'm not sure I have a good proposal if the class name is required but it's been generated and can't be found easily

**Comment by HaloFour** [comment]
IMO if there was any movement in this front it should be for top-level methods in general.  Extension methods would theoretically just fall out of that.

I also think that given that extensions are [being rethought](https://github.com/dotnet/csharplang/discussions/5496) in general it probably doesn't pay to devote new syntax specifically to a form that may end up being deprecated.

**Comment by Rekkonnect** [comment]
I'm a big fan of reducing boilerplate around extension methods. But there's a catch.

There is lying a proposal about roles and extensions (previously under the umbrella of "shapes"). This proposal keeps the declaration of a named type, but removes the `this` parameter and the `static` modifiers from every declared extension. For example (as the proposal currently stands):
```csharp
public extension BarExtensions : Bar
{
    public void Do() { /* ... */ }
}
```

In my honest opinion, I cannot disagree with your proposal, but I cannot judge yet because of the aforementioned open extension proposal. Adding both would feel like extensions are a dark area in the language and are bodged around. Additionally, consider the current scenarios with extension-dedicated classes:
- one small class containing just a few (no more than 3/4) functions, often times for enums
  - this current topic can be solved via #297
- one larger class containing a bulk of extensions for specific purposes, often times containing helper methods
  - top-level extensions would require helpers be declared either in the same file, making it very messy, or in another file, defeating the purpose of this proposal

Reducing nesting by one more space won't offer much to me anyway. It was great with the namespace because of the redundancy. This time, extensions suffer from greater problems, including *why* you have to use them, and *how much pain* declaring them is (`static` and `this` keywords all over the place).

**Reply by tebeco** [reply]
(haven't open/read both link yet on `public extensions ....` proposal
But then would that still be possible to have a `Top Level` approach
```cs
public extension ... Do(...) { code }
```
generates:
```cs
public extension ... 
{
  ...Do(...) { code }
}
```

given more or less restriction at build time on what can be generated or conflicts

**Reply by Rekkonnect** [reply]
```csharp
public extension ... Do(...) { code }
```

makes no sense. An `extension` declaration is still a (static) class(-like?) type. This means that you would have to declare things *inside* it, not *alongside* it.

Your proposal for shortening the process would yield
```csharp
public extension BarExtensions : Bar;

public void Do() { /* ... */ }
```

**Reply by tebeco** [reply]
does the "Type Name" matter, `BarExtensions` here that was my intial though
_"why do I have to create a container type everytime to write down an extension ?"_

I'm not implying it should replace the full verbose format, I'm saying maybe we don't need it.
that was easy to apply with the `(this Bar bar` with older syntax, could possibly create bloat with new one
